doozerlib/runtime.py

```diff
@@ -859,6 +859,45 @@ def git_clone(self, remote_url, target_dir, gitargs=None, set_env=None, timeout=
         cmd.append(target_dir)
         exectools.cmd_assert(cmd, retries=3, on_retry=["rm", "-rf", target_dir], set_env=set_env)
 
+    def lookup_source(self, meta):
```

Comment:
I think `detect_remote_source_branch` already did this pretty much?

Anyway, I want to throw a curveball here. I would expect we would do this after running (or at least initiating) the builds for a release. By that point, it's at least possible that the sources might already have drifted from what we built. Would we not want to capture what we built, rather than what is in the sources at some slightly later time?

We could do that by looking at the Dockerfiles in dist-git (via cgit). `doozer config:scan-sources` uses this logic already for comparing github sources with distgit contents.

Reply:
@sosiouxme There were a lot of code changes since this PR was open especially after Justin merged the fast rebase change. Last time when I visited the doozer code, it is not simple to reuse the existing functions and I didn't have enough confidence to modify them. Now I think I need to revisit that part of code and see if any parts can be reused.

Reply:
Capturing the actual builds is not covered in this PR. Capturing commit hashes after the initial build doesn't seem provide more value than recording commit hashes first and it will produces 2 different code paths.
I think the following can be a transitional route before we get a smart trigger:
- prepare-release job: records commit hashes of branch tips for all components. Triggers custom build job for them.
- Patch-release job: bump the commit hash to latest branch tip for an image, or specific a custom commit. Triggers custom build for any changes. 
- custom build job: will always build from the last recorded commit. Users usually don't need to directly trigger this job anyway unless explicitly rebuild without code change.
- build-sync job will always create an image stream with builds that matches the recorded hashes
@sosiouxme @jupierce would like to hear suggestions from you guys.